### ISSUE TITLE
Discrete vars in Metropolis sampler

### ISSUE THREAD

**Issue by hgbrian** [pymc-devs/pymc#655]
I think there is an issue with discrete vars in the Metropolis sampler.

The following code describes whether or not each of the vars being sampled is discrete
`self.discrete = np.array([v.dtype in discrete_types for v in vars])`
So the expectation is that you will have an array of Trues and Falses representing each var being sampled.

However, arraystep submits each var one at a time: i.e. it calls Metropolis([var])
Then for a discrete var, self.discrete looks like this:
`np.array([True])`

In the astep method of Metropolis, the following code is used:
`delta = np.atleast_1d(self.proposal_dist() * self.scaling)`
Here I’ll assume self.proposal_dist() returns a 1D array of size 3: e.g., a multinomial, [1,2,3].

Then the following code is called:
`delta[self.discrete] = round(delta[self.discrete], 0).astype(int)`
The intent is to round each discrete var in the list of vars.
However, if self.discrete is size 1 (one var), this does not work, and only the first element in the array is rounded.

Basically, the problem is that numpy pads boolean indexes with Falses.
I saw a comment suggesting this was undocumented behavior in numpy, but now I can't find it.

Here’s an example of 1D vs 2D indexing.

```
>>> discrete = np.array([True])
>>> x1d = np.array([1,2,3])
>>> x1d[discrete]
array([1])
>>> x2d = np.array([[1,2,3]])
>>> x2d[discrete]
array([[1, 2, 3]])`
```

What you really want is for x1d[[True]] to just return x1d, as though it was broadcast to x1d[[True,True,True]]…

I am sure there must be a better way to fix this (for example using np.atleast_2d almost works), but this is how I ended up fixing it:

```
if self.discrete.size == 1: # there is only one var, boolean indexing fails
    self.discrete = np.repeat(self.discrete, delta.shape[0])
```

**Comment by twiecki** [comment]
Hrm, that's an unfortunate state indeed. Thanks for the detailed explanation. Why does `np.atleast_2d` not work here?

**Comment by hgbrian** [comment]
Basically because delta has to be the same dimensions as q/q0.
`q = q0 + delta`
If q were always at least 2D too then it would work. 

When astep is called, it's called with 
`apoint = self.astep(bij.map(point), *inputs)`
which (I think) means it assumes it is sending the values/point from one var (q).

q is expected to be 1D since
`if (any(q_new != q0) or all(q0 == q)):`
fails for a q >1D. (any becomes ambiguous, but maybe np.array_equal would fix it)

It looks like there is a mismatch between the intent of vars (you can pass a list of vars to Metropolis and it will deal with each independently) and how q is handled (you pass Metropolis astep the values/"point" from one var, and not the values from a list of vars that happens to be length 1). 

Sorry I can't offer a neat solution... I don't know the history of these functions so I'm not sure whether you would want Metropolis astep to be more general and allow updating of several vars, or more clearly for updating one var, where delta[self.discrete] would not make sense within astep (because delta is the same dimensions as q, not the same dimensions as vars).

**Comment by twiecki** [comment]
The idea is certainly to allow updating of multiple variables. It is a bit weird to cast what should logically be 1-D array (let me know if you disagree here) to be 2-D arrays just get around an indexing artefact. Perhaps the cleanest solution would indeed be a special case logic for a special case behavior. What do you think?

Also would want to get @jsalvatier's input here.

**Comment by hgbrian** [comment]
The reason I thought atleast_2d might be a reasonable solution is because Metropolis is always passed a list of vars, even when there is only one var, so I thought it might be reasonable to always treat q as a list of arrays, one for each var. However, after looking at the code a bit more, I realize that all the qs are concatenated into a 1D array, so my assumption was incorrect.

Although my hack fixes the code for single vars (i.e., Metropolis([var]), which is how arraystep seems to be always called right now), I am not actually sure if the code around discrete should work at all...

proposal_dist is always 1D:

```
if S is None:
    S = np.ones(sum(v.dsize for v in vars))
self.proposal_dist = proposal_dist(S)
```

and so delta is always 1D too. 
`delta = np.atleast_1d(self.proposal_dist() * self.scaling)`
This makes sense since we're concatenating all the qs into one 1D array. 

However, it looks like self.discrete is indexing delta as though delta is a 2D array of vars:
`delta[self.discrete] = round(delta[self.discrete], 0).astype(int)`

So perhaps a better solution is something like (note I haven't tried this...):
`self.discrete = np.array([v.dtype in discrete_types for v in vars])`
changed to
`self.discrete = np.ravel([[v.dtype in discrete_types]*v.dsize for v in vars])`

(One final thing: vars is a built-in in Python)

**Comment by twiecki** [comment]
Yeah, I think `ravel` is the correct solution here. Thanks for figuring that out -- want to do a pull request?

**Comment by hgbrian** [comment]
Sounds good -- I'll do a couple of tests first.

**Comment by twiecki** [comment]
Great. Yeah, it would be good to have a unittest that reveals the current bug.

**Comment by jsalvatier** [comment]
Did this fix ever go in? It would be great to have this fixed.

**Comment by twiecki** [comment]
I don't think it did but should be pretty easy to apply.

**Comment by hvasbath** [comment]
Has this been resolved yet? I am currently facing a problem which might be related to this?
I have 2 discrete variables that basically resemble the indexes of a 2d array of size eg. 5x10.
I put the upper and lower bounds of these discrete variables to 0/4 and 0/9 respectively.
Still during sampling there are variables of value 10 or 5 being created and tried to sample, which results in an out of bound error- of course as these array indices do not exist. 

I can artificially circumvent this by subtracting 1 from each bound, but I guess this is not intended to be like this...
Best regards,
Hannes

**Comment by hvasbath** [comment]
I basically read the whole code the last day and know now what is going on.
It is basically free to sample past the bounds just afterwards it is being assigned negative infinite probability. Which would be ok for any problem that does not directly has to pick the Greensfunctions dependend on the model parameter. 
Anyways isnt it also very inefficient? It is trying to sample parts of the solution space where we know BEFORE that it is the worst you can do. Still it would run the whole forward model, which is the most expensive part of the calculation. Just to find AFTERWARDS ah yes it has negative infinite probability. I will try to fix that.

**Comment by fonnesbeck** [comment]
It is only inefficient for discrete variables, because continuous ones are automatically interval-transformed to the real line by PyMC 3. This allowed us to simplify the step method such that it did not have to worry about enforcing boundaries. Perhaps we need a `DiscreteMetropolis`, as we have in PyMC 2, that does not propose values not in the support of the variable?

**Comment by fonnesbeck** [comment]
Fixed via #979, I believe. Reopen if not.
